src/connectionpools.jl

```diff
@@ -0,0 +1,259 @@
+        # is released back to the Pod
+        while true
+            # this `wait` call will block on our Pod `lock` condition
+            # until a connection is `release`ed and the condition
+            # is notified
+            # println("connection pool maxxed out; waiting for connection to be released to the pod")
+            conn = wait(pod.lock)
+            if conn !== nothing
+                # println("checking recently released connection validity for reuse")
+                if isvalid(pod, conn)
+                    # println("connection just released to the Pod is valid and can be reused")
+                    conn.count += 1
+                    id = connectionid(conn.conn)
+                    # println("returning connection (id='$(id)')")
+                    pod.active[id] = conn
+                    return conn.conn
+                end
+            end
+            # if the Connection just returned to the Pod wasn't valid, the active
+            # count at least went down, so we should be able to create a new one
+            if length(pod.active) < pod.max
+                # println("connection just returned wasn't valid; creating new")
+                conn = ConnectionTracker(f())
+                id = connectionid(conn.conn)
+                # println("returning connection (id='$(id)')")
+                pod.active[id] = conn
+                return conn.conn
+            end
+            # If for some reason there were still too many active connections, let's
+            # start the loop back over waiting for idle connections to be returned
+            # Hey, we get it, writing threadsafe code can be hard
+        end
+    finally
+        unlock(pod.lock)
+    end
+end
+
+function release(pod::Pod{C}, conn::C; return_for_reuse::Bool=true) where {C}
+    lock(pod.lock)
+    try
+        # We first want to look up this connection object in our
+        # Pod `active` Dict that tracks active connections
+        id = connectionid(conn)
+        # if, for some reason, it's not in our `active` tracking Dict
+        # then something is wrong; you're trying to release a `Connection`
+        # that this Pod currently doesn't think is active
+        if !haskey(pod.active, id)
+            error("invalid connection pool release call; each acquired connection should be `release`ed ONLY once")
+        end
+        cp_conn = pod.active[id]
+        # remove the ConnectionTracker from our `active` Dict tracker
+        delete!(pod.active, id)
+        if return_for_reuse && isopen(conn)
+            # reset the idle timestamp of the ConnectionTracker
+            cp_conn.idle = time()
+            # check if there are any tasks waiting on a connection
+            if isempty(pod.lock)
+                # if not, we put the connection back in the pod idle queue
+                # println("returning connection (id='$(id)') to pod idle queue for reuse")
+                push!(pod.conns, cp_conn)
```

Comment:
With the scheme above, we'd just always do this `push!` here, and use `notify(pod.lock, nothing; all=false)`